cumulusci/utils/xml/metadata_tree.py

```diff
@@ -0,0 +1,265 @@
+from typing import Union, Generator
+
+from lxml import etree
+
+from .salesforce_encoding import serialize_xml_for_salesforce
+
+METADATA_NAMESPACE = "http://soap.sforce.com/2006/04/metadata"
+
+
+def parse(source):
+    """Parse a file by path or file object into a Metadata Tree
+
+    The parse() function supports any of the following sources:
+
+        * an open file object (make sure to open it in binary mode)
+        * a file-like object that has a .read(byte_count) method returning a byte string on each call
+        * a filename string or pathlib Path
+        * an HTTP or FTP URL string
+
+    Note that passing a filename or URL is usually faster than passing an open file or file-like object. However, the HTTP/FTP client in libxml2 is rather simple, so things like HTTP authentication require a dedicated URL request library, e.g. urllib2 or requests. These libraries usually provide a file-like object for the result that you can parse from while the response is streaming in.
+    """
+    if hasattr(source, "open"):  # for pathlib.Path objects
```

Comment:
While I generally like duck typing, this feels a bit too open (pun intended). For example if someone tries to pass in a ZipFile, this will try to call its open method and will raise a TypeError due to missing a required argument.